db_test.go

```diff
@@ -374,6 +375,156 @@ func (s *SurrealDBTestSuite) TestSmartUnMarshalQuery() {
+		vals := map[string]interface{}{}
+		for idx, user := range users {
+			query = append(query,
+				fmt.Sprintf("CREATE user_for_all SET id = $id_%d, Username = $user_%d, Password = $pass_%d;",
+					idx, idx, idx))
+			vals[fmt.Sprintf("id_%d", idx)] = user.ID
+			vals[fmt.Sprintf("user_%d", idx)] = user.Username
+			vals[fmt.Sprintf("pass_%d", idx)] = user.Password
+		}
+
+		data, err := s.db.Query(strings.Join(query, ""), vals)
+		s.Require().NoError(err)
+
+		result, err := surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+
+		// The result is ordered based on the server response, which is ordered
+		// by the record ID.
+		s.Equal("abcdef", result[0].Username)
+		s.Equal("ghijkl", result[1].Username)
+	})
+
+	s.Run("raw select query", func() {
+		data, err := s.db.Query("SELECT * FROM user_for_all", nil)
+		s.Require().NoError(err)
+
+		result, err := surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+
+		// The result is ordered based on the server response.
+		s.Equal("abcdef", result[0].Username)
+		s.Equal("ghijkl", result[1].Username)
+	})
+
+	s.Run("select query", func() {
+		data, err := s.db.Select(users[0].ID)
+		s.Require().NoError(err)
+
+		result, err := surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+
+		// The result is ordered based on the server response.
+		s.Equal("abcdef", result[0].Username)
+		s.Require().Len(result, 1) // Second item is not returned.
+	})
+
+	s.Run("select bulk query", func() {
+		data, err := s.db.Select("user_for_all")
+		s.Require().NoError(err)
+
+		result, err := surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+
+		// The result is ordered based on the server response.
+		s.Equal("abcdef", result[0].Username)
+		s.Equal("ghijkl", result[1].Username)
+	})
+
+	s.Run("delete record query", func() {
+		// Delete first user.
+		data, err := s.db.Delete(users[0].ID)
+		s.Require().NoError(err)
+
+		result, err := surrealdb.SmartUnmarshalAll[userForAll](data)
+
+		s.Require().NoError(err)
+		s.Empty(result[0])
+
+		// Double check that the second entry is still there.
+		data, err = s.db.Select("user_for_all")
+		s.Require().NoError(err)
+
+		result, err = surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+		s.Equal("ghijkl", result[0].Username) // abcdef was deleted above.
+
+		// Delete second user.
+		data, err = s.db.Delete(users[1].ID)
+		s.Require().NoError(err)
+
+		result, err = surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+		s.Empty(result[0])
+
+		// Double check that there is no entry.
+		data, err = s.db.Select("user_for_all")
+		s.Require().NoError(err)
+
+		result, err = surrealdb.SmartUnmarshalAll[userForAll](data)
+		s.Require().NoError(err)
+		s.Empty(result)
+	})
+
+	s.Run("use raw query with broken data", func() {
+		data, err := s.db.Query(`
+        CREATE user_dummy:xxxx SET Username = "x", Password = "xxxx";
```

Comment:
if you just create `users:xxxx` after tests it will clean up itself no need for manual cleanup

Reply:
I thought it would be still good to check the slice handling even after deletion. I could certainly change the name, but what do you think?

(On a separate note, my understanding is that a singular form is preferred, so it may be worth doing a clean up)
https://surrealdb.com/docs/surrealql/statements/relate

To be frank, all test cases using the same SurrealDB instance makes it really hard to track and test more complex scenarios. While it may not be the best to add extra dependencies, it's worth considering more extensive testing based on https://github.com/ory/dockertest like https://github.com/upsidr/surrealtest (I put this together for my testing), or use `surreal` CLI directly for in memory db run and execute as a part of each test case. I could put together an example for this in a separate PR?